### ISSUE TITLE
Process.wait and Process.wait2 not behaving as expected on Windows with JDK > 8

### ISSUE THREAD

**Issue by jcouball** [jruby/jruby#7515]
On Windows, Process.wait is not setting $? (or $CHILD_STATUS) and Process.wait2 is returning nil.

This is happening within my GitHub Actions. I created a simple test here that shows the behavior:
https://github.com/jcouball/process_spawn_test/actions/runs/3632082827

This workflow is a matrix build that runs the tests on two different JRuby and MRI versions on both Ubuntu and Windows. Only the JRuby builds on Windows fail.

The test code is very simple and can be found here:
https://github.com/jcouball/process_spawn_test/blob/main/spec/test_spec.rb

Duplicated here:

```ruby
require 'English'

describe 'Process#wait' do
  it 'should set the global $CHILD_STATUS variable' do
    pid = Process.spawn('exit 0')
    Process.wait(pid)
    expect($CHILD_STATUS).not_to be_nil
    expect($CHILD_STATUS.pid).to eq(pid)
  end
end

describe 'Process#wait2' do
  it 'should return a non-nil status' do
    pid = Process.spawn('exit 0')
    exited_pid, status = Process.wait2(pid)
    expect(status).not_to be_nil
    expect(status.pid).to eq(pid)
  end
end
```

I suspect this might be a configuration error with the Windows image I am using. I don't have access to a Windows computer or virtual environment so can't debug it on my own. Also, my Windows experience is rusty.

**Environment Information**
  
The environment where I am seeing unexpected behavior:

* JRuby 9.3.9.0 and JRuby 9.4.0.0
* Windows Server 2022 (20221127 update) (see [this page](https://github.com/actions/runner-images/releases/tag/win22%2F20221127.1) for more details of what is included in this Windows image)
* The following environment variables are included:
  * JAVA_OPTS: -Djdk.io.File.enableADS=true
  * JRUBY_OPTS: --debug -Xnative.verbose=true
* The Gemfile only includes the `rspec` gem

Environments where I see the expected behavior:

* JRuby 9.3.9.0 and JRuby 9.4.0.0 on Ubuntu
* MRI Rubys on both Ubuntu and Windows

**Expected Behavior**

* `Process.wait` should set $? / $CHILD_STATUS
* `Process.wait2` should return [pid, status] instead of nil.

**Actual Behavior**

* After calling `Process.wait`, $? / $CHILD_STATUS is nil.
* `Process.wait2` is returning nil instead of [pid, status].

**Comment by rdp** [comment]
I think I just ran into this as well.
Perhaps the pid values are wrong?  This snippet:

```
out = IO.popen("pause")
puts out.pid
```
gives crazy large numbers

> c:\ java --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED -jar jruby-complete-9.4.0.0.jar test2.rb
1801196366

with jdk 8 it's the right numbers like `15952`

**Comment by rdp** [comment]
Looks like it is getting this and then defaulting to hash code (unrelated suggestion: just raise instead of default to hashcode?)

```
java.lang.IllegalAccessException: class org.jruby.util.ShellLauncher$3 (in module org.jruby.dist) cannot access a member of class java.lang.ProcessImpl (in module java.base) with modifiers "private final"
        at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:392)
        at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:674)
        at java.base/java.lang.reflect.Field.checkAccess(Field.java:1102)
        at java.base/java.lang.reflect.Field.get(Field.java:423)
        at org.jruby.dist/org.jruby.util.ShellLauncher$3.getPid(ShellLauncher.java:745)
        at org.jruby.dist/org.jruby.util.ShellLauncher.reflectPidFromProcess(ShellLauncher.java:775)
        at org.jruby.dist/org.jruby.util.ShellLauncher.getPidFromProcess(ShellLauncher.java:656)
        at org.jruby.dist/org.jruby.util.ShellLauncher.runWithoutWait(ShellLauncher.java:644)
        at org.jruby.dist/org.jruby.util.ShellLauncher.runExternalWithoutWait(ShellLauncher.java:585)
        at org.jruby.dist/org.jruby.RubyProcess.spawn(RubyProcess.java:1779)
        at org.jruby.dist/org.jruby.RubyProcess$INVOKER$s$0$0$spawn.call(RubyProcess$INVOKER$s$0$0$spawn.gen)
        at org.jruby.dist/org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:824)
        at org.jruby.dist/org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:220)
        at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:372)
        at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:175)
        at bad_them.invokeOther0:spawn(bad_them.rb:1)
        at bad_them.RUBY$script(bad_them.rb:1)
        at bad_them.run(bad_them.rb)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
        at org.jruby.dist/org.jruby.ir.Compiler$1.load(Compiler.java:114)
        at org.jruby.dist/org.jruby.Ruby.runScript(Ruby.java:1277)
        at org.jruby.dist/org.jruby.Ruby.runNormally(Ruby.java:1194)
        at org.jruby.dist/org.jruby.Ruby.runNormally(Ruby.java:1176)
        at org.jruby.dist/org.jruby.Ruby.runNormally(Ruby.java:1212)
        at org.jruby.dist/org.jruby.Ruby.runFromMain(Ruby.java:991)
        at org.jruby.dist/org.jruby.Main.doRunFromMain(Main.java:398)
        at org.jruby.dist/org.jruby.Main.internalRun(Main.java:282)
        at org.jruby.dist/org.jruby.Main.run(Main.java:227)
        at org.jruby.dist/org.jruby.Main.main(Main.java:199)
```

This seems to workaround it but not sure if you want a more global fix, unit tests etc:

```
diff --git a/core/src/main/java/org/jruby/util/ShellLauncher.java b/core/src/main/java/org/jruby/util/ShellLauncher.java
index 17b861c..91e76f4 100644
--- a/core/src/main/java/org/jruby/util/ShellLauncher.java
+++ b/core/src/main/java/org/jruby/util/ShellLauncher.java
@@ -42,6 +42,8 @@ import java.io.PipedInputStream;
 import java.io.PipedOutputStream;
 import java.io.PrintStream;
 import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.nio.ByteBuffer;
 import java.nio.channels.FileChannel;
 import java.util.ArrayList;
@@ -745,7 +747,14 @@ public class ShellLauncher {
                         }

                     } catch (Exception e) {
-                        // ignore and use hashcode
+                        // JDK > 8 has a new way to look it up and also can't use "handle" field anymore
+                        try {
+                            Method pid = process.getClass().getSuperclass().getMethod("pid");
+                            Java.trySetAccessible(pid);
+                            return (long) pid.invoke(process);
+                        } catch (Exception e2) {
+                            // ignore and use hashcode
+                        }
                     }
                     return process.hashCode();
                 }
```

somewhat related: https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/issues/120

**Comment by headius** [comment]
This may be an issue running the complete jar with module support on more recent JDKs. There are various flags that need to be passed to open up the internals of the classes in question. Some earlier JDKs will just complain, but later ones will outright deny access.

I've added the "windows" label and will be getting a test environment set up soon to work on this and other issues.

**Comment by rdp** [comment]
I don't think it's an issue with the complete jar (it still fails with the --add-opens).  I think it is a windows thing on > JDK 8 yeah.

**Comment by byteit101** [comment]
Assigning myself to keep track of this for subspawn-win32

**Comment by jcouball** [comment]
@byteit101 and @headius: I was wondering what the plan is for this issue?

**Comment by byteit101** [comment]
The plan, as far as I'm aware, is for the next release (or the one after that) to ship subspawn-win32, however that is contingent on me getting access to a windows development environment where [master is actually green](https://github.com/jruby/jruby/issues/7621) so that I can validate that my changes actually work and don't break stuff.

**Comment by jcouball** [comment]
That sounds great. Forgive my ignorance, what is subspawn-win32? Is that a Win32 library? I search here and googled but didn't find anything obvious.

**Comment by byteit101** [comment]
It's a backend for my SubSpawn project: https://github.com/byteit101/subspawn/

JRuby currently uses SubSpawn for PTY.spawn and other PTY functions on Linux & macOS  as of 9.4

**Comment by jcouball** [comment]
Hey Patrick, wondering if progress is being made on this issue? JRuby support for my gem is blocked on this issue -- which isn't an emergency for me but important to some users of my gem. I noted that you have made quite a few changes on the subspawn project so thought you might have an update.

**Comment by byteit101** [comment]
Yeah! If you are lucky subspawn-win32 now works. You are generally lucky about 2/3 of the time. When you are unlucky, a read of a pipe from a dead process hangs indefinitely due to some native IO issue I haven't been able to figure out. @headius  offered to help me debug it, but he's been at conferences recently. If you don't need to read STDOUT/STDERR, though you can build master and use it without any problems, at least that I've come across. I've been unable to run it in the JRuby test suite  as that uses blocking STDOUT reads and thus hangs.

If you would like to try to help test, or solve the IO issue, either would be much appreciated. I was going to link you to the GHA builds, but I realized that I left them broken, and I'm about to head off on vacation for a week. Building master locally should be fine though, and you can also pop by #jruby on matrix as I'm usually there. I will be back and able to help again on the 21st.

I will mention that `` (backtick, or `%x{}`) is affected by this, but spawn + wait should be fine

**Comment by jcouball** [comment]
This behavior sounds familiar since that was exactly what I faced in my Ruby MRI implementation. However, I confess that I do not have much experience in Java (it's been about 20 years!). Unfortunately, I don't think I can be of much help.

I appreciate all your efforts and will continue to wait.

**Comment by rdp** [comment]
Seems working ok for me with that patch :) If you trust random builds: https://drive.google.com/file/d/1F523EobGRxITEWue3JF5BbhqF_XReWO1/view?usp=share_link

**Comment by jcouball** [comment]
When do you think this might be merged?

**Comment by headius** [comment]
Do we have a PR for this patch yet?

Sorry this fell off the priority list but let's wake it up and get it fixed for 9.4.5.0.

**Comment by byteit101** [comment]
That's a great question. This is waiting on debugging the windows hangs in subspawn-win32. I've been very busy this year, though things should be calming down for me in a few weeks in november.

**Comment by jcouball** [comment]
@headius and @byteit101 

❄️ Hello and Happy Holidays ❄️ 

Hoping to get an update as to when you think this might land in JRuby. I see this is targeted to 9.4.6.0. Is that still the plan?

I know you are busy folks so I am just trying to get your current thoughts.

**Comment by headius** [comment]
Currently nobody is working on the Windows support in subspawn, so this is waiting for some resource to step up.

If nobody else does it, I will try to do it myself, but it might be a little while. We would like to have this in next release.

**Comment by headius** [comment]
I have a Windows VM and will try to look at this for 9.4.9!

**Comment by headius** [comment]
This is actually simpler than it seems and I missed the details.

On Windows we do not dig out the PID properly, and instead return a hashcode. This is obviously not waitable, and for whatever reason we hang trying to wait for bogus PIDs on Windows.

Subspawn fixes this by actually implementing process management with FFI on Windows, returning a real PID. It is largely complete for common platforms other than Windows, but Windows needs some help due to https://github.com/byteit101/subspawn/issues/3. Once that's resolved and the rest of Windows support is cleaned up and completed, we can switch to Subspawn on all platforms and resolve this issue.

We probably will not make that move until JRuby 10, but we can update the optionally-enabled Subspawn in any JRuby 9.4 release. I think we should also patch the pid logic from ShellLauncher as described above.

**Comment by headius** [comment]
I have pushed a version of the patch from @rdp in #8310 but I am unable to test it due to us not having full FFI support on Win32 AArch64 yet.

**Comment by jcouball** [comment]
As of today, Sep 3, 2024, I am still seeing the unexpected behavior.

I have updated my tests to use the following JRuby versions: 9.3.15.0, 9.4.8.0, and head.

All still fail.

As a reminder, here is [the test code](https://github.com/jcouball/process_spawn_test/blob/main/spec/test_spec.rb).

and here is [the latest test run that failed](https://github.com/jcouball/process_spawn_test/actions/runs/10686421728/job/29621582782).

**Comment by jcouball** [comment]
Can this issue be reopened? Or do I have to create a new issue?

**Comment by headius** [comment]
@jcouball Sorry for delays on this... I'll see if we can do anything short term to improve this.

**Comment by headius** [comment]
@jcouball I pushed https://github.com/jruby/jruby/pull/8354 that may fix this issue for you. In that PR I use a Java 9+ API for accessing the PID so we can skip all the invasive reflection access. Would you be able to build and test that branch on your environment?

**Comment by headius** [comment]
I think I discovered why this fails: we don't actually open up `java.lang`, needed by the reflective access:

https://github.com/jruby/jruby/blob/9560c3a371d754d1066feff98c7ea71ba46cde23/bin/.jruby.module_opts

A workaround in the short term would be to edit this file in your local JRuby install, or modify the command line you use to start JRuby, and add the following flag:

```
--add-opens java.base/java.lang=org.jruby.dist
```

This requirement will be rendered moot once #8354 is merged.

**Comment by jcouball** [comment]
I don't have a windows environment. 

My test runs in a GitHub Action using the windows-latest platform and various JRuby versions installed by actions/setup-ruby.

I added the --add-opens flag to the JAVA_OPTS in [the workflow definition](https://github.com/jcouball/process_spawn_test/blob/main/.github/workflows/main.yml#L26). However, I get an error "Unrecognized option: --add-opens". See [the error in context here].(https://github.com/jcouball/process_spawn_test/actions/runs/11041869379/job/30673079120#step:3:41).

I am not sure if you were able to get a windows environment. If so, it should be trivial for you to run the script by cloning [the jcouball/process_spawn_test project](https://github.com/jcouball/process_spawn_test), and then running `bundle installed` followed by `rspec` with your newly built ruby in the path.

**Comment by headius** [comment]
@jcouball That line seems to indicate that it was running Java 8:

```
$ java -version
openjdk version "1.8.0_362"
OpenJDK Runtime Environment (Zulu 8.68.0.21-CA-macos-aarch64) (build 1.8.0_362-b09)
OpenJDK 64-Bit Server VM (Zulu 8.68.0.21-CA-macos-aarch64) (build 25.362-b09, mixed mode)
$ java --add-opens
Unrecognized option: --add-opens
```

Your issue was with Java > 8, so I think you may need to specify to use a Java version 9+ in your workflow to test that flag.

**Comment by headius** [comment]
Optimistically marking this as completed with the support for Java 9+ Process.pid... hopefully that is the last piece needed until we have full subspawn support.

**Comment by jcouball** [comment]
Still reproducing on JRuby 10.1.1.0 / JDK 21 — verified today (2026-08-08). Requesting reopen.

@headius you closed this in Oct 2024 as optimistically fixed by the Java 9+ `Process.pid()` support (#8354), with the note *"hopefully that is the last piece needed until we have full subspawn support."* Unfortunately it wasn't — the behavior is unchanged nearly two years and two major versions later, and I've now confirmed the `--add-opens` theory is not the cause.

### Environment

```
jruby 10.1.1.0 (4.0.0) 2026-07-22 aaa95d57e1 OpenJDK 64-Bit Server VM 21.0.12+8-LTS on 21.0.12+8-LTS +indy +jit [x86_64-mswin32]
```

* GitHub Actions `windows-latest`
* Temurin JDK 21.0.12 (via `actions/setup-java`)
* JRuby installed by `ruby/setup-ruby` with `ruby-version: jruby` (resolves to latest stable)

### Reproducer (no gems required)

```ruby
require 'English'

pid = Process.spawn('ruby', '-e', 'exit 0')
Process.wait(pid)
puts "wait  : $CHILD_STATUS = #{$CHILD_STATUS.inspect}"

pid = Process.spawn('ruby', '-e', 'exit 0')
puts "wait2 : #{Process.wait2(pid).inspect}"

pid = Process.spawn('ruby', '-e', 'exit 42')
Process.wait(pid)
puts "exit42: $CHILD_STATUS = #{$CHILD_STATUS.inspect}"
```

Observed on JRuby/Windows:

```
wait  : $CHILD_STATUS = nil
wait2 : nil
exit42: $CHILD_STATUS = nil
```

Note that `Process.wait2` returns `nil` outright rather than a `[pid, status]` pair with a nil
status. Neither `spawn` nor `wait` raises — the child does run and exit; only the status is lost.
The failure is not specific to a zero exit code.

On MRI (Linux, macOS, Windows) and on JRuby/Linux, all three lines report a proper
`Process::Status`.

### The `--add-opens` workaround does not help on 10.1.1.0

You previously suggested `--add-opens java.base/java.lang=org.jruby.dist`, since
`bin/.jruby.module_opts` didn't open `java.lang` for the reflective access. I ran a four-way
matrix on `windows-latest` / JDK 21 / JRuby 10.1.1.0 against the script above:

| Variant | `$CHILD_STATUS` after `wait` | `wait2` |
|---|---|---|
| baseline, no flag | `nil` | `nil` |
| `JAVA_OPTS=--add-opens=java.base/java.lang=org.jruby.dist` | `nil` | `nil` |
| `JRUBY_OPTS=-J--add-opens=java.base/java.lang=org.jruby.dist` | `nil` | `nil` |
| `JAVA_OPTS=--add-opens=java.base/java.lang=ALL-UNNAMED` | `nil` | `nil` |

All four are identical. Run: https://github.com/main-branch/process_executer/actions/runs/31286086810

Also worth noting — the shipped `.jruby.module_opts` in 10.1.1.0 no longer mentions `java.lang`
at all:

```
--add-opens java.base/java.io=org.jruby.dist
--add-opens java.base/java.nio.channels=org.jruby.dist
--add-opens java.base/sun.nio.ch=org.jruby.dist
--add-opens java.management/sun.management=org.jruby.dist
```

So whatever is dropping the status now, it doesn't appear to be the module-access problem
diagnosed in 2024.

### Standing regression test

I keep a standalone two-example spec plus a manually-dispatchable workflow so this can be
re-checked against any JRuby release:

* https://github.com/main-branch/process_executer/tree/main/process_spawn_test
* https://github.com/main-branch/process_executer/actions/workflows/process-spawn-test.yml

Most recent matrix run — {MRI, JRuby} × {ubuntu-latest, windows-latest}, only `jruby / windows-latest`
fails: https://github.com/main-branch/process_executer/actions/runs/31285362980

Happy to run any further probe you'd like on that workflow; I don't have a local Windows
environment, but the turnaround on a dispatch is a couple of minutes.

### Impact

This blocks JRuby/Windows support in [process_executer](https://github.com/main-branch/process_executer)
(tracked as main-branch/process_executer#156). Without a subprocess status the library can't detect
command failure or enforce timeouts, so we currently document JRuby on Windows as unsupported and
limit our JRuby support policy to Linux.

**Comment by byteit101** [comment]
@jcouball 
>  until we have full subspawn support.

In theory I was supposed to be that half of the conditional statement, but I got very busy and I don't remember the status. Can you test the RC version of the subspawn gem on windows and see if it works for you?

In theory I have more time to explore the final subspawn push this year

**Comment by jcouball** [comment]
@headius @byteit101 I have a local Windows box now, so I could finally test this myself
instead of guessing from CI logs. The short version:

* #8354 fixed the PID. The remaining bug is that `Process.wait` doesn't wait.
* SubSpawn master fixes that on Windows. My spec suite goes from 5 failures to 5 passes.
* The SubSpawn RC on RubyGems can't be installed on JRuby at all, and would recurse
  forever on Windows if it could.
* Two other bugs in master keep it from being a drop-in replacement, one on Windows and
  one on POSIX. Both are small.

Everything below was measured on:

```
jruby 10.1.1.0 (4.0.0) 2026-07-22 aaa95d57e1 OpenJDK 64-Bit Server VM 21.0.12+8-LTS
  on 21.0.12+8-LTS +indy +jit [x86_64-mswin32]
Windows 11 Pro 26200, Zulu JDK 21.0.12, stock jruby-bin-10.1.1.0.zip and its own launcher
```

## `Process.wait` never waits

#8354 worked. Process.spawn returns the child's real PID, which I checked by having
the child report its own Process.pid back. So the bad-PID explanation from 2024 no
longer holds, and the bug is still here without it.

The remaining failure isn't a bad status coming back. It's that `wait` doesn't wait.

`child.rb` sleeps 2 seconds, writes a file, then exits 3. Spawn it, time the wait, and
check whether the child has done anything yet:

```
run 1: wait -> nil in 0.0042s   $?=nil   child file present after wait? false
        child file present 3s later? true  "child pid=12952 done"
run 2: wait -> nil in 0.0001s   $?=nil   child file present after wait? false
        child file present 3s later? true  "child pid=18168 done"
run 3: wait -> nil in 0.0002s   $?=nil   child file present after wait? false
        child file present 3s later? true  "child pid=4612 done"
```

It returns `nil`, doesn't block, and doesn't raise `Errno::ECHILD`. The child runs to
completion regardless, and reports the same PID that `spawn` handed back. A GitHub Actions
`windows-latest` runner agrees: an example that spawns `ruby -e 'sleep 2'` and asserts the
wait took at least 1.5 seconds measures 0.00013 seconds there.

I got 0 non-nil results in 45 trials across three child shapes, `exit 0`, a child that
writes to stdout, and a child that sleeps, and 0 again on a second run of the same 45.

One caveat. Very early on, before I settled on the stock launcher, I saw `Process.wait`
return a pid and a correct status exactly once. I have not reproduced it since, including
about a hundred later attempts and eight aimed specifically at recreating the conditions.
Take it as a hint that this might be a race rather than a hard never, but I can't give you
a reproducer for the working case.

The rest of the API splits in two:

| API | JRuby 10.1.1.0 / Windows | MRI 3.4.2, same machine |
| --- | --- | --- |
| `Process.wait(pid)` | `nil`, `$?` nil | pid + status |
| `Process.waitpid(pid)` | `nil`, `$?` nil | pid + status |
| `Process.waitpid2(pid)` | `nil` | `[pid, status]` |
| `Process.wait2(pid)` | `nil` | `[pid, status]` |
| `Process.waitall` | raises `NotImplementedError` | `[[pid, status]]` |
| `Process.detach(pid).value` | blocks, but see below | `exit 3` |
| `system(...)` | correct `$?`, `exit 7` for `exit 7` | same |
| backticks | output plus correct `$?` | same |
| `Open3.capture2/3` | correct status | same |
| `IO.popen` then `close` | correct `$?` | same |

Building a `Process::Status`, decoding exit codes, and reaping all work on Windows. Only
the `waitpid` family is broken. Everything that holds onto the launcher's own process
handle is fine.

`Process.waitall` says why:

```
NotImplementedError: wait unsupported or native support failed to load;
  see https://github.com/jruby/jruby/wiki/Native-Libraries
  waitall at org/jruby/RubyProcess.java:1138
```

Two smaller findings.

`Process.detach(pid).value` blocks, but misreads the status. A child that exits 3 comes
back as `#<Process::Status: pid 8756 signal 3>` with `exitstatus` nil. MRI reports `exit 3`
for the same child on the same machine. `system` and `Open3` decode that child correctly,
so it's only this path.

`Process.wait` returning `nil` instead of raising `Errno::ECHILD` is its own spec
deviation, and it's why this is so quiet in libraries. Nothing raises. The status just
disappears.

The `--add-opens` theory is dead, and I checked it two ways. I reproduced the bug with the
stock launcher, then again with `jruby.jar` on the classpath and every package from
`bin/.jruby.module_opts` re-targeted to `ALL-UNNAMED`, plus `java.base/java.lang`. Same
result both times. The jnr-posix "modules may not be open" warning doesn't appear in the
runs that still fail, so access isn't what's missing.

## SubSpawn results

@byteit101, I hit two problems before I could get to your actual question. Then a good
answer, then two more problems.

### The published RC won't install on JRuby

```
$ jgem install subspawn --pre
Building native extensions. This could take a while...
ERROR:  Error installing subspawn:
        ERROR: Failed to build gem native extension.
...
make failed
```

`engine-hacks` 0.2.0.pre1 was only ever pushed as a `ruby` platform gem carrying the C
extension. Your gemspec has the branch for the other case,

```ruby
if RUBY_PLATFORM =~ /java/
  spec.platform = "java"
else
  spec.platform = Gem::Platform::RUBY
  spec.extensions = ["ext/engine_hacks/extconf.rb"]
end
```

but no `java` platform build was published, so RubyGems picks the `ruby` one and tries to
compile a C extension under JRuby. This isn't Windows specific. It's any JRuby.

Building that gemspec with JRuby produces `engine-hacks-0.2.0.pre1-java.gem`. After
installing that by hand, the other four gems installed cleanly.

### The published RC then recurses forever on Windows

```
Error: Your application used more stack memory than the safety cap of 2048K.
  subspawn/replace-builtin.rb:41  ->  subspawn.rb:326  ->  replace-builtin.rb:41  ->  ...
```

Two guards for the same question disagree. `subspawn.rb:308` in the RC asks

```ruby
if SubSpawn::Platform.method_defined? :waitpid2
```

which checks instance methods, but `SubSpawn::Win32.waitpid2` is a singleton method. So
that guard is false, and `SubSpawn.wait` takes the fallback branch and calls
`Process.wait`. Meanwhile `replace-builtin.rb:39` asks
`SubSpawn::Platform.method(:waitpid2)`, which does find the singleton method, so it has
already replaced `Process.wait` with `SubSpawn.wait`. Round and round. Measured on Win32:

```
SubSpawn::Platform            = SubSpawn::Win32  (Class)
P.method_defined?(:waitpid2)  = false   <- guard in subspawn.rb:308
P.respond_to?(:waitpid2)      = true
P.method(:waitpid2) truthy?   = true    <- guard in replace-builtin.rb:39
```

The recursion is Windows only, because `subspawn-posix` doesn't define `waitpid2` at all.
POSIX breaks on the same line a different way, further down.

You already fixed the recursion. master has `respond_to?` at `subspawn.rb:311`.

### master fixes the status bug

I built the five gemspecs from master with JRuby 10.1.1.0, installed them, and required
`subspawn/replace-builtin`:

```
SubSpawn::Platform = SubSpawn::Win32
wait  : -> 12504  $CHILD_STATUS = #<Process::Status: pid 12504 exit 0>
wait2 : [2568, #<Process::Status: pid 2568 exit 0>]
exit42: exitstatus = 42
slow child (2s, exit 3): waited 2.73s, status=#<Process::Status: pid 1540 exit 3>
Process.wait non-nil 20/20
```

Right status, right exit code, and it actually blocks. My spec suite goes from 5 failures
to 5 passes on the same machine and the same Ruby, with only the backend switched.

So the status bug is fixed. Two other problems stop master from dropping straight in, and
I only found them after putting it through CI.

### subspawn-win32 picks the wrong file off PATH

The same spec suite that passes locally fails on a GitHub Actions Windows runner, at
`spawn` instead of at `wait`:

```
Failure/Error: pid = Process.spawn('ruby', '-e', 'exit 0')
SystemCallError:
  Unknown error (SystemCallError) 193 - Spawn Error: CreateProcess: {win: 193, c: 0}
```

Windows error 193 is `ERROR_BAD_EXE_FORMAT`. JRuby's `bin` ships both `ruby`, a shell
script with no extension, and `ruby.bat`. When `ruby/setup-ruby` puts that directory on
PATH, SubSpawn resolves `ruby` to the extensionless script and hands it to `CreateProcess`,
which can't run it. It doesn't appear to consult `PATHEXT`.

I reproduced it locally by putting JRuby's `bin` first on PATH, which is the only
difference from my passing run:

```
backend: stock
  spawn('ruby') -> pid 12984
backend: subspawn
  spawn('ruby') -> SystemCallError: Spawn Error: CreateProcess: {win: 0, c: 0}
```

Stock JRuby launches it fine. SubSpawn also spawns a `.bat` fine when handed the path
directly, so PATH resolution is the part that's wrong. The reported error code is
inconsistent too, 193 on the runner and 0 locally for the same failure.

### `subspawn/replace-builtin` doesn't load at all on POSIX

On Linux the require raises before any example runs:

```
NameError: undefined method 'waitpid2' for class '#<Class:0x93ce24d>'
  replace-builtin.rb:39
```

Same line as the recursion bug, failing the other way. `Object#method` raises `NameError`
when the method doesn't exist rather than returning nil, so
`if SubSpawn::Platform.method(:waitpid2)` can never be false. On Win32 it returns a Method
and the guard is wrongly true. On POSIX, where `waitpid2` isn't defined, it raises.
`respond_to?` fixes both, the same way it did in `subspawn.rb`.

This matters beyond my test setup, because `subspawn/replace-builtin` is what JRuby's own
`USE_SUBSPAWN=true` requires on non-Windows platforms.

### I couldn't reproduce the win32 IO hang

byteit101/subspawn#3 is what you said was blocking you, so I hammered on it under master:

```
  Open3.capture2 100B       30/30 ok
  Open3.capture3 100B       30/30 ok
  capture2 1MB              5/5 ok
  capture3 1MB both pipes   5/5 ok
```

70 pipe reads plus 1 MB through both pipes at once, no hangs. That's a modest sample
against something you put at roughly 1 in 3, so I won't call it fixed. It doesn't look
like the same failure mode is still there, though. Tell me what shape of IO used to hang
for you and I'll run that exact case.

### `IO.popen(array)` raises in master

```ruby
IO.popen(['ruby', 'emit.rb', '100'])
# ArgumentError: First argument must be an pair TODO: check this
```

`SubSpawn.__compat_parser` reads a leading array as the `[cmdname, argv0]` pair and
rejects it when the length isn't 2. In MRI, `IO.popen(array)` and `Process.spawn(array)`
mean argv. The varargs form works fine, and stock JRuby handles the array form correctly.

## JRuby refuses `USE_SUBSPAWN=true` on Windows

Whenever JRuby ships this, the built-in opt-in won't help the platform that needs it.
`Ruby.java` logs `env USE_SUBSPAWN=true is unsupported on Windows at this time` and loads
nothing. I required `subspawn/replace-builtin` directly to get around that for testing.

## What would unblock this

1. Change `replace-builtin.rb:39` to `respond_to?`. It's the same fix you
   already made in `subspawn.rb`. Right now that line breaks Win32 one way and POSIX the
   other way.
2. Make SubSpawn's PATH lookup honor `PATHEXT` on Windows, so `spawn('ruby')` finds
   `ruby.bat` instead of the extensionless `ruby` script.
3. @byteit101 publishes a `java` platform build of `engine-hacks`. Without it, SubSpawn
   0.2.x can't be installed on JRuby on any platform.
4. A new prerelease cut from master with a bumped version. master is still labelled
   0.2.0.pre1, so the fixed code and the broken published gem share a version number.
5. The `IO.popen(array)` case, though that one didn't block my testing.

@headius, if the `waitpid` family keeps its current implementation for a while, the working
half of that table is a usable workaround. `system`, backticks, `Open3`, and `IO.popen` all
return correct statuses on JRuby/Windows today. I can route around this in my own gem with
those, so this stopped being urgent for me.

## Unrelated, spotted along the way

JRuby on Windows drops embedded double quotes from argv. Same behavior with and without
SubSpawn, so this isn't a SubSpawn thing:

```
                   sent                   got (JRuby)     got (MRI)
embedded dquote    "a\"b"                 ["ab"]          ["a\"b"]
ruby snippet       "puts \"x\" * 3"       ["puts x * 3"]  ["puts \"x\" * 3"]
```

Say the word and I'll open a separate issue rather than bury it here.

## Standing regression test

The workflow now runs `{MRI, JRuby} x {ubuntu-latest, windows-latest} x {stock, subspawn}`,
so this can be rechecked against any JRuby release or SubSpawn commit:

* https://github.com/main-branch/process_executer/tree/main/process_spawn_test
* https://github.com/main-branch/process_executer/actions/workflows/process-spawn-test.yml

The run that turned up the two master bugs is
https://github.com/main-branch/process_executer/actions/runs/32317593229. The
`jruby / windows-latest / stock` job fails there for the reason this issue is about. The
two `subspawn` jobs fail for the PATHEXT and `respond_to?` reasons, not the status bug.

Happy to run anything else on the local box too. Turnaround is seconds now, and I can get
you a thread dump if something hangs.

**Comment by byteit101** [comment]
@jcouball thanks, I will try to push through the rc to a release and then promote it into what JRuby uses. I'm hoping to have some time the next few weekends to look at this

**Comment by byteit101** [comment]
Just a quick update: I sat down to look at this last weekend, and was able to reproduce https://github.com/byteit101/subspawn/issues/3 about half the time. I will hopefully get a more complete setup uploaded this weekend, and dig into it more.
